### ISSUE TITLE
Integrate validation module with interface

### ISSUE THREAD

**Issue by webberwang** [codelab-ai/builder#1760]
An interface is not complete without validation.

https://stackoverflow.com/questions/66009560/yup-vs-joi-for-frontend-validation#:~:text=Joi%20is%20more%20popular%20for,times%20smaller%20than%20Joi%20%2D%20link

~~I'm thinking we could encode this library as a separate Validation model, which we can connect to an Interface model.~~

A interface type is actually close related to the validation, for example if we choose an integer/number type, the validation is associated with that.

If we wanted a date validation, we would add a Date to our TypeKind.

But some are shared to all types such as required, nullable.

We'll choose a handful of important rules first that apply to all types

- [ ] Required
- [ ] Nullable

**Comment by webberwang** [comment]
To re-iterate, we'll integrate a validation library under the hood, the values from the column such as required/nullable will be passed into an instantiated class such as 

## An Example

```
string()
  .min(3, 'must be at least 3 characters long')
  .max(6, 'must be less than 3 characters long')
  .email('must be a valid email')
  .validate('no'); // ValidationError
```

In this example above, we could imagine a `User` interface with `email` field set to `String` TypeKind.

`min`, `max`, `isEmail` column are added, and each cell contains 2 fields.

1 example
`min` - an integer field and a string field to support the custom error message

## Example Improvement

The above example, we wouldn't actually use a `String` `TypeKind`, but rather create a new `TypeKind` of `Email` type, then we wouldn't need `isEmail`

The above 2 field per cell is also not currently supported by our interface builder, perhaps we could do.

`min` & `min.message` to support 2 fields?

- [ ] How to support 2 fields per cell?

**Comment by mohasarc** [comment]
I think it makes perfect sense to implement validation. There are a few things that confused me though:

So to my understanding an `InterfaceType` would include multiple fields, these can be of any `PrimitiveType` `TypeKind` or an `InterfaceType` `TypeKind`. We want to have some validation for these fields so that we make sure the interface is being used correctly. What I don’t understand is the `column`s and `cell`s, are these UI elements, or code blocks? And how are they related to fields and interfaces, hierarchy-wise?

Additionally I have a few more questions:

1. Where do we use interfaces in our applications, I know that atoms have interfaces, are there other places?
2. When setting the value of an atom’s property (interface field), we already have some validation based on the type string/number (as you mentioned their validation is associated with them), here I think we want to add more validation rules like
`Nullable` etc..?
3. `TypeKinds` are currently hard-coded, are they gonna be dynamic. Additionally, you mentioned string `TypeKind`, did you mean `PrimitiveType`?
4. If `TypeKinds` can be dynamically added, isn't that equivalent to creating a new interface?
5. How can the validation be used for the required atom props issue for example, I think while creating them we’d assign some validation for these props?
6. How the user experience would look like, I imagine: The user would create an interface just like what we have now, but additionally, there will be multiple other inputs per type filed. For example, a `string` type, might have `min`, `max`, and `Nullable` inputs. is this more or less similar to what you have in mind?

**Comment by webberwang** [comment]
So these validation rules would apply to an `InterfaceType` when we assign the `PrimitiveType`.

![image](https://user-images.githubusercontent.com/26264925/187439405-04751603-c174-4fe5-b09d-02540108e8c3.png)

Depending on the `PrimitiveType`, we may have different columns. Perhaps we would need a validation cell that opens up a modal. Since we don't know which rules apply until the field is assigned a type.

1. A Store can also assign an Interface to a Store key, this is our local state management
2. Fields like `nullable` would apply to all, we would have to go over the api and see which ones apply to which primitive
3. I meant `PrimitiveType`
4. It wouldn't be dynamic but hard code new types to `TypeKind`, for example adding email
5. As the props go into the atom, we could find the associated interface (& its validation rules), we can add a check to see whether this is required or not and perform the proper action
6. Currently I think a single column `Validation` would contain a button to open a modal for these rules, this still needs some thought

**Comment by mohasarc** [comment]
I think it make sense to start looking at the database and how to represent `Validation` there first.

Looking at the current schema, we can see that there are multiple `TypeKind`s and some of them have an `INTERFACE_FIELD` relationship between them like `InterfaceType` with `PrimitiveType`, `ActionType`, `AppType`, and `PageType`, or an `ARRAY_ITEM` relationship like `ArrayType` with `PrimitiveType`

![Screenshot from 2022-09-02 10-21-21](https://user-images.githubusercontent.com/51242349/188082699-28f5e97a-f3d1-44d4-ba7f-8c6c8e0045d9.png)

However, I think there are some missing relationships, because we seem to be able to have any `TypeKind` as an `InterfaceType`'s field. for example `ArrayType` here: 

![Screenshot from 2022-09-02 10-18-33](https://user-images.githubusercontent.com/51242349/188081035-430b4d1d-fc0a-434c-a0f5-12a49772a536.png)

So I think the previous schema should have a few more relationships and probably look something like this:

![Screenshot from 2022-09-02 10-35-00](https://user-images.githubusercontent.com/51242349/188085926-3c0a5462-4116-4b91-90bd-1d92e075c647.png)

On the other hand, regarding `ValidationRules`, I think all `TypeKind`s can be validated with some validation rule. Implementation-wise, I think each `ValidationRule` can have its own node and there can be a `VALIDATE_WITH` relationship between a `TypeKind` and a `ValidationRule`. I'll try to highlight this using the following simplified schema:

![CodelabTypeValidation-simplified_schema (2)](https://user-images.githubusercontent.com/51242349/188088611-9ff80072-38b2-4201-93e9-4eba84c09e15.jpg)

Note I entroduced an abstract node just to simplify the graph, basically whatever happens on the abstract node happens on its instances.

In this graph we can notice that any `TypeKind` can be `VALIDATED_WITH` any `ValidationRule`; an `InterfaceType` can have an `INTERFACE_FIELD` of any `TypeKind`; and, `ArrayType` can have `ARRAY_ITEM` of any `TypeKind`.

Having in mind your suggestion regarding  the support for function on slack, I have also added `Input` and `Output` nodes to `Action/FuncitonType`, which can be on any `TypeKind` type again. This might be one way to implement function support.

To put this chema in a concrete example, I'll use the User example you use earlier. Using that we can have the following graph:

![CodelabTypeValidation-usecase_1](https://user-images.githubusercontent.com/51242349/188094556-3056437a-a136-4ae6-9ced-ea7f9fe9fe6e.jpg)
 let me know what you think about this approach.

Draw.io file: 
[CodelabTypeValidation.zip](https://github.com/codelab-ai/builder/files/9476190/CodelabTypeValidation.zip)

**Comment by webberwang** [comment]
Wow this is a very impressive breakdown and I understand completely what you mean.

To address the first issue of missing `INTERFACE_FIELD`, that is because we only added fields that we've used and initially we didn't use that many. But yes an interface field could be any time.

Secondly, I do think each validation rule should be it's own node. This way reusability is maximized as we can compose based on each type.

---

Our initial version could hard code these rules, but eventually I think we'd want to abstract them with a Function/Action type so we can add these functionalities via the UI.

It would be similar to an Atom type, but instead a Validation type. Perhaps related to https://codelabai.slack.com/archives/C01TALGNKMZ/p1662013544004989

This last part I haven't fully worked out, but lets hardcode them first.

**Comment by mohasarc** [comment]
Great that we're on the same page regarding the DB design! I think next step would be to think in terms of code, where and how validation will be implemented and how `Yup` will be integrated. My initial though is to have `ValidationRule` model and use `Yup` inside it. I'll look into it in more details.

----------------
I'm not sure I unserstood what is being referred in this idea. Do you mena hard code validation rules for now, and later use `Function/ActionType` to add validations to the DB?

**Comment by webberwang** [comment]
Nevermind about that second part, I'd imagine we need some sort of dynamic rule chain https://stackoverflow.com/questions/52941791/yup-dynamic-validation-chain

## Option 1

Create a long `when` chain that contains all the combinations of rules, this is when all rules are used. Then user can pass in data via context to see which ones are used.

## Option 2

Use `eval` to dynamically call the function

`string().default('pat').required(),`

```
```

**Comment by mohasarc** [comment]
After doing some research I found a library that helps creating dynamic yup schema (https://github.com/kristianmandrup/schema-to-yup. It seems promising especially that it uses a standard JSON Schema layout (http://json-schema.org). The only downside is that there are no types available for this library.

On the other hand, regarding implementation, I think we need to:

- Create a new `ValidationRule` DB node for each rule. And make a relation between it and `TypeKind` as shown in the diagram above
- Create a new MobX model `ValidationRule`
- Have a new field in `BaseType` for an array of `ValidationRules`
- Assign validation rules to types while creating/modifiying an interface.
- Execute the validation rule in `react-hook-form`'s `Controller` component using `rules` property. Note we use `Controller` to wrap AntD controlled components, where we'll be recieving the user's input to set the value for a particular field.

**Comment by webberwang** [comment]
Should we consider https://ajv.js.org/ if we're going the JSON schema route, seems much more native with this library.

The MobX model is a nice touch, keeps our logic more contained. 

We would likely couple this with a custom form that we build for our users

**Comment by mohasarc** [comment]
Yeah I was just checking ajv it seems like a nice library. I'll have a more detailed look into it.

Do you mean the forms users build inside the apps they create using our platform? if so, the validation I think should be applied to the `onChange` or `onSubmit` callback's params or return type. And yes they can create their custom forms on the platform level. I think the only missing piece would be the feedback or error message we throw when the input is invalid. 

Another idea, is just to let the users validate their components using native AntD or MUI validation parameter by passing a custom validaton function.

**Comment by antonmihaylov** [comment]
I think ajv is a good option to consider. I'm not sure if it's still there, but at some point we had some code that transformed an Interface to json schema. 
The validation schema @mohasarc  made looks very good and flexible. The ValidationRules could be supplied to the json schema transformer and included in their json-schema form in the final schema. Then ajv can do the hard work and validate it.

**Comment by antonmihaylov** [comment]
By accident, I came across this old plan for adding validations when we created the type system. 

It is similar, but the difference seems to be where the validators are. Here they are connected to the fields, whereas in the current plan they are on the types

<img width="798" alt="image" src="https://user-images.githubusercontent.com/57956282/189063383-570311bc-0f65-447e-ba88-a0ba9f0d26a7.png">

That makes types more reusable - you don't have to create `String` and `String, required` type. But it is more limiting since you can only validate interfaces.

**Comment by webberwang** [comment]
Would we have a case where we need to validate non-interfaces?

**Comment by antonmihaylov** [comment]
Can't see a need currently, Interfaces are the root types

**Comment by mohasarc** [comment]
Please correct me on this, from what I understand, a field is any `TypeKind` that is connected to `InterfaceType` with an `INTERFACE_FIELD` relashionship. In that case we can apply the validation only on this subset of `TypeKind`s.

**Comment by webberwang** [comment]
I think it's just semantics here, we validate interfaces, but the implementation is through the fields

**Comment by mohasarc** [comment]
Now I think I understand the motivatiom behind having validationRules linked to interfaces. I've updated the schema accordingly and this is the result:

![CodelabInterfaceFieldValidation-simplified_schema](https://user-images.githubusercontent.com/51242349/190708183-862920a4-47f8-4524-aca5-425c678b4d7a.jpg)

Because JSON Schema supports primitive types as well as arrays and because each type have a different set of rules I linked the `ValidationRule` to them. Additionally, because multiple interfaces can be connected to the same PrimitiveType or array, I linked `ValidationRule` to `InterfaceType`. This way each `ValidationRule` is unique per interface field.

A `ValidationRule` in this case have a few properties namely, `name`, `value`, `errorMessage`, and `description`. Another option, would be to build the schema in the frontend while adding/modifiying the interface field, and store it as a schema object. The second option is more flexible if we wanted to write custom schema for a specific field, but I think the first might be easier to implement.

Here is a usecase:
![CodelabInterfaceFieldValidation-usecase_1](https://user-images.githubusercontent.com/51242349/190708789-b9f357b0-eaa4-4eab-89e0-198e88d27959.jpg)

In terms of implementation in the frontend, It would be the same as before: there would be a `ValidationRule` model that is linked to the field object in case the field is of a primitive type or Array type.

I'm thinking of leaving the array validation for now, and implement it after primitive types validation is complete.

Draw.io file:
[CodelabInterfaceFieldValidation.zip](https://github.com/codelab-app/builder/files/9586525/CodelabInterfaceFieldValidation.zip)

**Comment by webberwang** [comment]
We wouldn't need to link validation rule to the primitive types themselves (in the database) right, since it can be inferred through interface type

**Comment by mohasarc** [comment]
Actually now that I think about it we need to somehow link the validation rules to the field itself, but the problem is the field is a relationship here. The reason is that one interface can have multiple fields of the same type and each field can have multiple validation rules. Linking the validation rules to the interface and the primitive type wouldn't be enough when the interface has multiple fields of the same type.

**Comment by mohasarc** [comment]
What do you think of changing the `INTERFACE_FIELD` relationship into a node. this way we'll need to link a validation rule with only the field node.

**Comment by webberwang** [comment]
The validation rule could be keyed by `field.key` though

**Comment by webberwang** [comment]
I think we would link a single validation rule to both the interface and the field's type, this way we know which interface it belongs to, and which type it references.

**Comment by webberwang** [comment]
I think turning `INTERFACE_FIELD` into a node is an interesting solution, would solve the edge reference issue

**Comment by mohasarc** [comment]
The only concern of turning the `INTERFACE_FIELD` into a node would be migrating the current database into the new schema, which I don't think is a big deal at this stage.

**Comment by antonmihaylov** [comment]
That's how it used to be, do you remember why we changed it? Was it just for simplicity or was there another reason?

**Comment by webberwang** [comment]
> That's how it used to be, do you remember why we changed it? Was it just for simplicity or was there another reason?

I can't remember if Field was ever a node before?

**Comment by antonmihaylov** [comment]
> > That's how it used to be, do you remember why we changed it? Was it just for simplicity or was there another reason?
> 
> 
> 
> I can't remember if Field was ever a node before?
> 

It was at some point. Maybe it was before neo4j

**Comment by webberwang** [comment]
> 

Yeah for Dgraph I think, only cause they couldn't have a model on an edge i believe

**Comment by antonmihaylov** [comment]
Most likely yes. Is putting the validation in the edge attributes an option wirth considering?

**Comment by webberwang** [comment]
That's possible, but we're mixing 2 different models together. Validation and Field would be the same model.

Making a Field a node has other benefits, like upserting on edge which isn't supported in the GraphQL api by default

**Comment by antonmihaylov** [comment]
They don't have to be in the same object, if we can put an array of ValidationRules there.

True, but it has drawbacks too, like having to refactor half of thr type module

**Comment by webberwang** [comment]
On second thought this does seem good since Validation has access to the interface and the field type

```
  interface Field @relationshipProperties {
    id: ID!
    key: String!
    name: String
    description: String
    validation: ValidationRule
  }
```

**Comment by mohasarc** [comment]
I think having validations in the edge attributes makes perfect sense and it would be the easiest and quickest solution. The only limitation would be if we wanted to apply validations on primitive array items, but we can solve this by allowing the `value` field of `ValidationRule` to accept an object or an array of objects to accomodate some schema keywords for array type like https://ajv.js.org/json-schema.html#items
